Implement mapping of HttpResponseMessage to CachedResponse to fix #152

hampos opened the pull request:
## Closes #152 
- #152 

I have created a test that exposes the deserialization problem from issue #152 where HttpResponse cannot be deserialized from cache.
I used the  `InMemoryJsonHandle` from the pull request https://github.com/TomPallister/Ocelot/pull/151 in order to do this.

I opened the pull request on develop, I think you can change this if you create another branch for this issue.

Commit:
changed name to cache options to fix issue #146

hampos committed:
Add acceptance test that exposes JSON deserialization bug from issue #146

- Create InMemoryJsonHandle for CacheManager that mimics DictionaryHandle but uses ICacheSerializer to serialize/deserialize values instead of saving references
- Add CacheManager.Serialization.Json package
- Add StartupWithCustomCacheHandle class that extends Startup and overrides ConfigureServices to register InMemoryJsonHandle
- Add GivenOcelotIsRunningUsingConsulToStoreConfigAndJsonSerializedCache method to initiate Ocelot with StartupWithCustomCacheHandle
- Add test should_return_response_200_with_simple_url_when_using_jsonserialized_cache

hampos committed:
Create Acceptance test that exposes issue #152

- Add GivenOcelotIsRunningUsingJsonSerializedCache() that initializes Ocelot with InMemoryJsonHandle
- Add should_return_cached_response_when_using_jsonserialized_cache test

TomPallister commented:
@hampos looks good! Thanks for the help.

hampos committed:
Change Consul port to 9502 on should_return_response_200_with_simple_url_when_using_jsonserialized_cache() test

hampos committed:
Merge branch 'feature/fix-#146' into feature/fix-#152

hampos committed:
Merge branch 'develop' into feature/fix-#146

# Conflicts:
#	test/Ocelot.AcceptanceTests/ConfigurationInConsulTests.cs
#	test/Ocelot.AcceptanceTests/Steps.cs

- Move StartupWithCustomCacheHandle to Startup.cs

hampos committed:
Merge branch 'feature/fix-#146' into feature/fix-#152

# Conflicts:
#	test/Ocelot.AcceptanceTests/Steps.cs

hampos committed:
Implement mapping of HttpResponseMessage to CachedResponse to enable distributed caching

- Add CachedResponse class that holds HttpResponse data
- Add mapping methods in OutputCacheMiddleware to create HttpResponseMessage from CachedResponse and vice versa
- Replace HttpResponseMessage with CachedResponse in services registrations
- Replace HttpResponseMessage with CachedResponse in OutputCacheController's IOcelotCache

hampos committed:
Fix unit tests for OutputCacheMiddleware and OutputCacheController by replacing HttpResponseMessage with CachedResponse

hampos committed:
Merge branch 'develop' into feature/fix-#152

# Conflicts:
#	src/Ocelot/DependencyInjection/ServiceCollectionExtensions.cs
#	test/Ocelot.AcceptanceTests/ConfigurationInConsulTests.cs

TomPallister commented on src/Ocelot/Cache/CachedResponse.cs:
```diff
@@ -0,0 +1,21 @@
+			Body = "";
+		}
+
+		public HttpStatusCode StatusCode { get; set; }
```
Can we have private setters here please, the project only has public setters for the file configuration stuff because microsofts framework for config cant support private setters.

hampos replied:
yep, sure no problem!

TomPallister commented on src/Ocelot/Cache/Middleware/OutputCacheMiddleware.cs:
```diff
@@ -65,11 +67,54 @@ public class OutputCacheMiddleware : OcelotMiddleware
+
+		internal Task<HttpResponseMessage> CreateHttpResponseMessage(CachedResponse cached)
+		{
+			return Task.Run(() =>
```
Why are we spinning up a task here? I dont think we need the overhead

hampos replied:
Yes you're right, I originally put the Task in the signature to be uniform with CreateCachedResponse, then thought to go premature on it and spinned a task because it does a Base64 conversion.
Should I drop the Task completely? Or return a Task.FromResult() for uniformity?

TomPallister replied:
I would just drop the task!

TomPallister commented on test/Ocelot.AcceptanceTests/Startup.cs:
```diff
@@ -43,5 +44,24 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
-    }
+	}
+
+	public class StartupWithCustomCacheHandle : Startup
```
This is cool I should do this more!!

TomPallister commented on src/Ocelot/Cache/Middleware/OutputCacheMiddleware.cs:
```diff
@@ -47,7 +48,8 @@ public class OutputCacheMiddleware : OcelotMiddleware
                 _logger.LogDebug("cache entry exists for {downstreamUrlKey}", downstreamUrlKey);
 
-                SetHttpResponseMessageThisRequest(cached);
+				var response = await CreateHttpResponseMessage(cached);
```
spacing seems a little off! Might just be github

hampos replied:
No it's not github, I probably have different settings on my visual studio and messes things up.
What are your settings on TextEditor -> C# -> Tabs?

hampos replied:
Maybe we can add a .editorconfig to enforce specific identation settings on the solution?

TomPallister replied:
Yeh good idea!

TomPallister replied:
i use vscode for Ocelot so not sure what I'm using...tbh im a shitty developer who doesnt care about tabs and spaces :E

TomPallister replied:
Do you have anything you would recommend for a default .editorconfig?

TomPallister reviewed (changes requested):
Thank you very much for the contribution! Looks good. Please can you check my comments and see what you think?

hampos committed:
Add .editorconfig with default identation settings (spaces with size 4)

hampos committed:
Re-format broken files with new identation settings

hampos committed:
Add Startup_WithConsul_And_CustomCacheHandle class

- Use Startup_WithConsul_And_CustomCacheHandle in GivenOcelotIsRunningUsingConsulToStoreConfigAndJsonSerializedCache step

hampos committed:
Do minor cleanups

- Rename StartupWithCustomCacheHandle to Startup_WithCustomCacheHandle for better readability
- Remove cachemanager settings Action in Startup since it is not used anymore

hampos committed:
Drop Task in CreateHttpResponseMessage - unnecessary overhead

hampos committed:
Make setters private in CachedResponse

- Rework CreateCachedResponse to use new CachedResponse constructor

TomPallister commented:
@hampos looks good. Are you happy to merge it?

hampos commented:
Yes if you're ok I'm good for the merge!
